neps/nep-0413.md

```diff
@@ -0,0 +1,153 @@
+  callbackUrl?: string; // Applicable to browser wallets (e.g. MyNearWallet). This is the callback url once the signing is approved. Defaults to `window.location.href`.
+  meta?: string; // Applicable to browser wallets (e.g. MyNearWallet) extra data that will be passed to the callback url once the signing is approved.
+}
+```
+
+### Structure and Signature
+In order to respect all the points layed in [rationales](#rationale-and-alternatives), `verifyOwner` must:
+
+1. Embed the passed `message` into a predefined structure (see below).
+2. Stringify the structure 
+2. Compute the `SHA256 hash` of the *stringified structure*.
+3. Use a full-access key to sign the resulting `SHA256 hash`.
+
+#### Structure
+The structure must have the following interface:
+
+```rust
+struct Payload {
+  accountId: string; // Mandatory: the account name as plain text (e.g. "alice.near")
+  message: string; // The same message passed in `VerifyOwnerParams.message` 
+  blockId: string; // The hash of a block created close to the time of signature
+  publicKey: string; // public counterpart of the key used to sign, encoded as a string with format "<key-type>:<base-64-key-bytes>"
+}
+```
+
+#### Signature
+In order to create a signature, the `Payload` must be converted into its `string` representation (respecting the attribute's order given above), and then hashed using `SHA256`.
+
+For example, assuming that:
+1. `alice.near` wants to sign the message `"berryclub.io"`.
+2. The wallet stores a private key which public counterpart is: `ed25519:6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y`
+3. A recent block was produced with the block hash `12FaoFyjYGmuxXnvWGYqxvUqGesSjdFVBNWXkXvoj41M`.
+
+The wallet should sign the **`SHA256` hash** of the following string:
+
+```jsx
+sha256.hash(`{"accountId":"alice.near","message":"berryclub.io","blockId":"12FaoFyjYGmuxXnvWGYqxvUqGesSjdFVBNWXkXvoj41M","publicKey":"ed25519:6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y"}`)
+```
+
+### Output Interface
+`verifyOwner` must return a object containing the **base64** representation of the `signature`, and all the data necessary to verify such signature. 
+
+```jsx
+interface AuthenticationToken {
+  accountId: string; // The account name as plain text (e.g. "alice.near")
+  message: string; // The same message passed in `VerifyOwnerParams.message` 
+  blockId: string; // The hash of a block created close to the time of signature
+  publicKey: string; // The public counterpart of the key used to sign, expressed as a string with format "<key-type>:<base-64-key-bytes>"
+  signature: string; // The base64 representation of the signature.
+}
+```
+
+### Returning the signature
+Web Wallets, such as [MyNearWallet](https://mynearwallet.com), should directly return the `AuthenticationToken` to the `VerifyOwnerParams.callbackUrl`, passing each attribute as a URL fragment. For example: `<callbackUrl>?accountId=<accountId>&message=<message>...`.
```

Comment:
The suggestion is to use fragments but the example uses query string parameters. I'm assuming at a minimum we want the `signature` behind a `#` but the rest are probably OK, e.g.
```
<callbackUrl>?accountId=<accountId>&message=<message>...#<signature>
```

Reply:
Fixed, thanks!